ext/drx/scatter_gather_aarch64.c

```diff
@@ -38,6 +38,481 @@
+            EMIT(eor_sve_pred_b, opnd_create_reg_element_vector(scratch_pred, OPSZ_1),
+                 opnd_create_predicate_reg(scratch_pred, false),
+                 opnd_create_reg_element_vector(sg_info->mask_reg, OPSZ_1),
+                 opnd_create_reg_element_vector(scratch_pred, OPSZ_1));
+
+            /* cpy      gather_dst_reg.elsz, scratch_pred/m, #0, lsl #0 */
+            EMIT(cpy_sve_shift_pred,
+                 opnd_create_reg_element_vector(sg_info->gather_dst_reg,
+                                                sg_info->element_size),
+                 opnd_create_predicate_reg(scratch_pred, true), OPND_CREATE_INT8(0),
+                 opnd_create_immed_uint(0, OPSZ_1b));
+        } else {
+            /* We don't care about any values in the dst register so zero the whole thing.
+             */
+
+            /* dup      gather_dst_reg.elsz, #0, lsl #0 */
+            EMIT(dup_sve_shift,
+                 opnd_create_reg_element_vector(sg_info->gather_dst_reg,
+                                                sg_info->element_size),
+                 OPND_CREATE_INT8(0), opnd_create_immed_uint(0, OPSZ_1b));
+        }
+    }
+
+    /* pfalse   scratch_pred.b */
+    EMIT(pfalse_sve, opnd_create_reg_element_vector(scratch_pred, OPSZ_1));
+
+    instr_t *loop_label = INSTR_CREATE_label(drcontext);
+    instr_t *end_label = INSTR_CREATE_label(drcontext);
+
+    instrlist_meta_preinsert(bb, sg_instr, loop_label);
+
+    /* pnext scratch_pred.elsz, mask_reg, scratch_pred.elsz */
+    EMIT(pnext_sve, opnd_create_reg_element_vector(scratch_pred, sg_info->element_size),
+         opnd_create_reg(sg_info->mask_reg));
+
+    /* b.none   end */
+    instrlist_preinsert(
+        bb, sg_instr,
+        INSTR_XL8(INSTR_PRED(INSTR_CREATE_bcond(drcontext, opnd_create_instr(end_label)),
+                             DR_PRED_SVE_NONE),
+                  orig_app_pc));
+
+    /* lastb    scratch_gpr, scratch_pred, index_reg.elsz */
+    EMIT(lastb_sve_scalar, opnd_create_reg(scratch_gpr), opnd_create_reg(scratch_pred),
+         opnd_create_reg_element_vector(sg_info->index_reg, sg_info->element_size));
+
+    if (sg_info->is_load) {
+        /* ldr[bh]  scratch_gpr, [base_reg, scratch_gpr, mod #amount] */
+        opnd_t mem = opnd_create_base_disp_shift_aarch64(
+            sg_info->base_reg, scratch_gpr, sg_info->extend, sg_info->scaled, 0, 0,
+            sg_info->scalar_value_size, sg_info->extend_amount);
+
+        if (sg_info->is_scalar_value_signed) {
+            const reg_id_t ld_dst =
+                reg_resize_to_opsz(scratch_gpr, sg_info->element_size);
+            switch (sg_info->scalar_value_size) {
+            case OPSZ_1: EMIT(ldrsb, opnd_create_reg(ld_dst), mem); break;
+            case OPSZ_2: EMIT(ldrsh, opnd_create_reg(ld_dst), mem); break;
+            case OPSZ_4: EMIT(ldrsw, opnd_create_reg(ld_dst), mem); break;
+            default: DR_ASSERT_MSG(false, "Invalid scatter_gather_info_t data");
```

Comment:
No 8-byte version for signed load?

Reply:
The signed loads sign extend the loaded value to the destination register size (either a 32-bit W register or 64-bit X register). It only makes sense for loads that are smaller than the register size. There is no `ldrsd` instruction (nor a `ld1sd` vector equivalent) because there are no bits to extend when you are loading a 64-bit value to a 64-bit register.